projects/packages/identity-crisis/src/class-url-secret.php

```diff
@@ -0,0 +1,162 @@
+ */
+
+namespace Automattic\Jetpack\IdentityCrisis;
+
+use Jetpack_Options;
+use WP_Error;
+
+/**
+ * IDC URL secret functionality.
+ */
+class URL_Secret {
+
+	/**
+	 * The options key used to store the secret.
+	 */
+	const OPTION_KEY = 'identity_crisis_url_secret';
+
+	/**
+	 * Secret lifespan (5 minutes)
+	 */
+	const LIFESPAN = 300;
+
+	/**
+	 * The URL secret string.
+	 *
+	 * @var string|null
+	 */
+	private $secret = null;
+
+	/**
+	 * The URL secret expiration date in unix timestamp.
+	 *
+	 * @var string|null
+	 */
+	private $expires_at = null;
+
+	/**
+	 * Initialize the class.
+	 */
+	public function __construct() {
+		$secret_data = $this->fetch();
+
+		if ( $secret_data !== null ) {
+			$this->secret     = $secret_data['secret'];
+			$this->expires_at = $secret_data['expires_at'];
+		}
+	}
+
+	/**
+	 * Fetch the URL secret from the database.
+	 *
+	 * @return array|null
+	 */
+	private function fetch() {
+		$data = Jetpack_Options::get_option( static::OPTION_KEY );
+
+		if ( $data === false || empty( $data['secret'] ) || empty( $data['expires_at'] ) ) {
+			return null;
+		}
+
+		if ( time() > $data['expires_at'] ) {
+			Jetpack_Options::delete_option( static::OPTION_KEY );
+			return null;
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Create new secret and save it in the options.
+	 *
+	 * @throws Exception Thrown if unable to save the new secret.
+	 *
+	 * @return bool
+	 */
+	public function create() {
+		$secret_data = array(
+			'secret'     => $this->generate_secret(),
+			'expires_at' => time() + static::LIFESPAN,
+		);
+
+		$result = Jetpack_Options::update_option( static::OPTION_KEY, $secret_data );
+
+		if ( ! $result ) {
+			throw new Exception( esc_html__( 'Unable to save new URL secret', 'jetpack-idc' ), 'unable_to_save_url_secret' );
+		}
+
+		$this->secret     = $secret_data['secret'];
+		$this->expires_at = $secret_data['expires_at'];
+
+		return true;
+	}
+
+	/**
+	 * Get the URL secret.
+	 *
+	 * @return string|null
+	 */
+	public function get_secret() {
+		return $this->secret;
+	}
+
+	/**
+	 * Get the URL secret expiration date.
+	 *
+	 * @return string|null
+	 */
+	public function get_expires_at() {
+		return $this->expires_at;
+	}
+
+	/**
+	 * Check if the secret exists.
+	 *
+	 * @return bool
+	 */
+	public function exists() {
+		return $this->secret && $this->expires_at;
+	}
+
+	/**
+	 * Adds `url_secret` to the `jetpack.idcUrlValidation` URL validation endpoint.
+	 * Adds `url_secret_error` in case of an error.
+	 *
+	 * @param array $response The endpoint response that we're modifying.
+	 *
+	 * @return array
+	 * phpcs:ignore Squiz.Commenting.FunctionCommentThrowTag -- The exception is being caught, false positive.
+	 */
+	public static function add_secret_to_url_validation_response( array $response ) {
+		try {
+			$secret = new URL_Secret();
+
+			if ( ! $secret->create() ) {
+				throw new Exception( esc_html__( 'Unable to create an URL secret.', 'jetpack-idc' ), 'unable_to_create_url_secret' );
```

Comment:
I think we could make this cleaner by avoiding this approach of throwing Exception inside a try/catch block.
Also, optionally we could consider that if we fail to generate the secret we could just return the original `$response` without the error

Reply:
The idea is to log the error in Kibana, we don't need to throw it for the user.
We could instead save it into Tracks/Hue, but it would be a bit harder to lookup.

And the idea behind throwing the exception is to only handle those exceptions in a single place (the `catch` block) instead of copying the error handling.

Reply:
I'm fine with returning the error but I strongly suggest we don't throw inside the `try/catch` block. How about something like:

```
$secret = new URL_Secret();

try {
    $secret->create();
 }
 catch ( Exception $e ) {
    $response['url_secret_error'] = new WP_Error( $e->getCode(), $e->getMessage() );
}

if ( $secret->exists() ) {
	$response['url_secret'] = $secret->get_secret();
}

return $response;
```

Reply:
Removed: https://github.com/Automattic/jetpack/pull/32005/commits/b65f1919756900972c93eb9fa6c6cb6b2938dfb0